homeassistant/components/bluetooth/active_update_processor.py

```diff
@@ -0,0 +1,88 @@
+    BluetoothChange,
+    BluetoothServiceInfoBleak,
+)
+from .passive_update_processor import (
+    PassiveBluetoothDataProcessor,
+    PassiveBluetoothDataUpdate,
+)
+from homeassistant.core import HomeAssistant, callback
+
+_T = TypeVar("_T")
+
+
+class ActiveDataProcessor(PassiveBluetoothDataProcessor):
+    """A DataProcessor that collects data passively from advertisements but can also poll."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        update_method: Callable[
+            [BluetoothServiceInfoBleak], PassiveBluetoothDataUpdate[_T]
+        ],
+        needs_poll_method: Callable[[BluetoothServiceInfoBleak, float | None], bool],
+        poll_method: Callable[
+            [BluetoothServiceInfoBleak],
+            Coroutine[Any, Any, PassiveBluetoothDataUpdate[_T]],
+        ],
+    ) -> None:
+        """Initialize the processor."""
+        super().__init__(update_method)
+
+        self._hass = hass
+        self._needs_poll_method = needs_poll_method
+        self._poll_method = poll_method
+        self._last_poll: float | None = None
+        self._poll_limiter = asyncio.Semaphore(2)
+
+    def needs_poll(self, service_info: BluetoothServiceInfoBleak) -> bool:
+        """Return true if time to try and poll."""
+        poll_age: float | None = None
+        if self._last_poll:
+            poll_age = time.monotonic() - self._last_poll
+        return self._needs_poll_method(service_info, poll_age)
+
+    async def async_poll(self, service_info: BluetoothServiceInfoBleak) -> None:
+        """Poll the device to retrieve any extra data."""
+        # There is already 1 poll running and 1 poll waiting
+        if self._poll_limiter.locked():
+            return
+
+        async with self._poll_limiter:
+            # A poll finished while we were waiting - so do we still need to poll
+            # at all??
+            if not self.needs_poll(service_info):
+                return
+
+            new_data = await self._poll_method(service_info)
+
+            self.devices.update(new_data.devices)
+            self.entity_descriptions.update(new_data.entity_descriptions)
+            self.entity_data.update(new_data.entity_data)
+            self.entity_names.update(new_data.entity_names)
+            self.async_update_listeners(new_data)
+
+            self._last_poll = time.monotonic()
+
+    @callback
+    def async_handle_bluetooth_event(
+        self,
+        service_info: BluetoothServiceInfoBleak,
+        change: BluetoothChange,
+    ) -> None:
+        """Handle a Bluetooth event."""
+        super().async_handle_bluetooth_event(service_info, change)
+
+        # See if its time to poll
+        # We use bluetooth events to trigger the poll so that we scan as soon as
+        # possible after a device comes online or back in range, if a poll is due
+        if self.needs_poll(service_info):
+            self._hass.add_job(self.async_poll(service_info))
```

Comment:
We're inside the event loop so we should use the async api, not `hass.add_job`. Use `hass.async_create_task` to schedule a coroutine.